In [26]:
#  Ebnable HTML/CSS 
from IPython.core.display import HTML
HTML("<link href='https://fonts.googleapis.com/css?family=Passion+One' rel='stylesheet' type='text/css'><style>div.attn { font-family: 'Helvetica Neue'; font-size: 30px; line-height: 40px; color: #FFFFFF; text-align: center; margin: 30px 0; border-width: 10px 0; border-style: solid; border-color: #5AAAAA; padding: 30px 0; background-color: #DDDDFF; }hr { border: 0; background-color: #ffffff; border-top: 1px solid black; }hr.major { border-top: 10px solid #5AAA5A; }hr.minor { border: none; background-color: #ffffff; border-top: 5px dotted #CC3333; }div.bubble { width: 65%; padding: 20px; background: #DDDDDD; border-radius: 15px; margin: 0 auto; font-style: italic; color: #f00; }em { color: #AAA; }div.c1{visibility:hidden;margin:0;height:0;}div.note{color:red;}</style>")

___
Enter Team Member Names here (double click to edit):

- Hari Sanadhya
- Mustafa Sakarwala
- Alfeo Sabay

________

# In Class Assignment Three
In the following assignment you will be asked to fill in python code and derivations for a number of different problems. Please read all instructions carefully and turn in the rendered notebook (or HTML of the rendered notebook)  before the end of class.

<a id="top"></a>
## Contents
* <a href="#Loading">Loading the Data</a>
* <a href="#distance">Measuring Distances</a>
* <a href="#KNN">K-Nearest Neighbors</a>
* <a href="#naive">Naive Bayes</a>

________________________________________________________________________________________________________
<a id="Loading"></a>
<a href="#top">Back to Top</a>
## Downloading the Document Data
Please run the following code to read in the "20 newsgroups" dataset from sklearn's data loading module.

In [2]:
from sklearn.datasets import fetch_20newsgroups_vectorized
import numpy as np
from __future__ import print_function

# this takes about 30 seconds to compute, read the next section while this downloads
ds = fetch_20newsgroups_vectorized(subset='train')

# this holds the continuous feature data (which is tfidf)
print('features shape:', ds.data.shape) # there are ~11000 instances and ~130k features per instance
print('target shape:', ds.target.shape) 
print('range of target:', np.min(ds.target),np.max(ds.target))
print('Data type is', type(ds.data), float(ds.data.nnz)/(ds.data.shape[0]*ds.data.shape[1])*100, '% of the data is non-zero')

features shape: (11314, 130107)
target shape: (11314,)
range of target: 0 19
Data type is <class 'scipy.sparse.csr.csr_matrix'> 0.1214353154362896 % of the data is non-zero


## Understanding the Dataset
Look at the description for the 20 newsgroups dataset at http://qwone.com/~jason/20Newsgroups/. You have just downloaded the "vectorized" version of the dataset, which means all the words inside the articles have gone through a transformation that binned them into 130 thousand features related to the words in them.  

**Question Set 1**:
- How many instances are in the dataset? 
- What does each instance represent? 
- How many classes are in the dataset and what does each class represent?
- Would you expect a classifier trained on this data would generalize to documents written in the past week? Why or why not?
- Is the data represented as a sparse or dense matrix?

####Answers
* Number of instances in the dataset: 11314
* Document
* 20
* If it's literature content it would not change much and the model could work. However if the content was technology, medicine or politics there will be new words not used before. It depedends on the type of document.
* Sparse Data





___
<a id="distance"></a>
<a href="#top">Back to Top</a>
## Measures of Distance
In the following block of code, we isolate three instances from the dataset. The instance "`a`" is from the group *computer graphics*, "`b`" is from from the group *recreation autos*, and "`c`" is from group *recreation motorcycle*. **Exercise for part 2**: Calculate the: 
- (1) Euclidean distance
- (2) Cosine distance 
- (3) Jaccard similarity 

between each pair of instances using the imported functions below. Remember that the Jaccard similarity is only for binary valued vectors, so convert vectors to binary using a threshold. 

**Question for part 2**: Which distance seems more appropriate to use for this data? **Why**?

In [7]:
from scipy.spatial.distance import cosine
from scipy.spatial.distance import euclidean
from scipy.spatial.distance import jaccard
import numpy as np

# get first instance (comp)
idx = 550
a = ds.data[idx].todense()
a_class = ds.target_names[ds.target[idx]]
print('Instance A is from class', a_class)

# get second instance (autos)
idx = 4000
b = ds.data[idx].todense()
b_class = ds.target_names[ds.target[idx]]
print('Instance B is from class', b_class)

# get third instance (motorcycle)
idx = 7000
c = ds.data[idx].todense()
c_class = ds.target_names[ds.target[idx]]
print('Instance C is from class', c_class)

# Enter distance comparison below for each pair of vectors:
print('Euclidean Distance\n ab:', euclidean(a,b), 'ac:', euclidean(a,c), 'bc:', euclidean(b,c))
print('Cosine Distance\n ab:', cosine(a,b), 'ac:', cosine(a,c), 'bc:', cosine(b,c))
print('Jaccard Dissimilarity (vectors should be boolean values)\n ab:', jaccard(a>0,b>0), 'ac:', jaccard(a>0,c>0), 'bc:', jaccard(b>0,c>0))

print('\n\nThe most appropraite distance is...') 
print('Cosine. It clearly delineates between topics effectively. B and C are closest (by far),',)
print('and A is about the same distance from B and C. Euclidean also works well, but not quite as good as "cosine"')


Instance A is from class comp.graphics
Instance B is from class rec.autos
Instance C is from class rec.motorcycles
Euclidean Distance
 ab: 1.0985184671870858 ac: 1.1891405425398234 bc: 0.9177794226661624
Cosine Distance
 ab: 0.603371411376 ac: 0.707027614956 bc: 0.421159534335
Jaccard Dissimilarity (vectors should be boolean values)
 ab: 0.882113821138 ac: 0.875471698113 bc: 0.908794788274


The most appropraite distance is...
Cosine. It clearly delineates between topics effectively. B and C are closest (by far),
and A is about the same distance from B and C. Euclidean also works well, but not quite as good as "cosine"


___
# Start of Live Session Assignment
___
<a id="KNN"></a>
<a href="#top">Back to Top</a>
## Using scikit-learn with KNN
Now let's use stratified cross validation with a holdout set to train a KNN model in `scikit-learn`. Use the example below to train a KNN classifier. The documentation for `KNeighborsClassifier` is here: http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html  

**Exercise for part 3**: Use the code below to test what value of `n_neighbors` works best for the given data. *Note: do NOT change the metric to be anything other than `'euclidean'`. Other distance functions are not optimized for the amount of data we are working with.* 

**Question for part 3**: What is the accuracy of the best classifier you can create for this data (by changing only the `n_neighbors` parameter)? 

In [9]:
from sklearn.cross_validation import StratifiedShuffleSplit
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from IPython.html import widgets 

cv = StratifiedShuffleSplit(ds.target, n_iter = 1, test_size = 0.5, train_size=0.5)

# fill in the training and testing data and save as separate variables
for trainidx, testidx in cv:
    # note that these are sparse matrices
    X_train = ds.data[trainidx] 
    X_test = ds.data[testidx] 
    y_train = ds.target[trainidx]
    y_test = ds.target[testidx]

    
# fill in your code  here to train and test

# calculate the accuracy and print it for various values of K
for K in range(1,20):
    clf = KNeighborsClassifier(n_neighbors=K, weights='uniform', metric='euclidean')
    clf.fit(X_train,y_train)
    prediction = clf.predict(X_test)
    acc = accuracy_score(y_test, prediction)
    print('Accuracy of classifier with %d neighbors is: %.2f'%(K,acc))


Accuracy of classifier with 1 neighbors is: 0.59
Accuracy of classifier with 2 neighbors is: 0.54
Accuracy of classifier with 3 neighbors is: 0.52
Accuracy of classifier with 4 neighbors is: 0.51
Accuracy of classifier with 5 neighbors is: 0.49
Accuracy of classifier with 6 neighbors is: 0.48
Accuracy of classifier with 7 neighbors is: 0.47
Accuracy of classifier with 8 neighbors is: 0.46
Accuracy of classifier with 9 neighbors is: 0.46
Accuracy of classifier with 10 neighbors is: 0.45
Accuracy of classifier with 11 neighbors is: 0.44
Accuracy of classifier with 12 neighbors is: 0.43
Accuracy of classifier with 13 neighbors is: 0.43
Accuracy of classifier with 14 neighbors is: 0.42
Accuracy of classifier with 15 neighbors is: 0.42
Accuracy of classifier with 16 neighbors is: 0.41
Accuracy of classifier with 17 neighbors is: 0.41
Accuracy of classifier with 18 neighbors is: 0.40
Accuracy of classifier with 19 neighbors is: 0.40


The best accuracy is .59 with 1 neighbors.

**Question for part 3**: With sparse data, does the use of a KDTree representation make sense? Why or Why not?

Enter your answer below:

KDTree representation does not make sense with sparse data because most of the data points will be on one side. For example in this case we're only evaluating less than 1% of the data. 


_____
## KNN extensions - Centroids
Now lets look at a very closely related classifier to KNN, called nearest centroid. In this classifier (which is more appropriate for big data scenarios and sparse data), the training step is used to calculate the centroids for each class. These centroids are saved. Unknown attributes, at prediction time, only need to have distances calculated for each saved centroid, drastically decreasing the time required for a prediction. 

**Exercise for part 4**: Use the template code below to create a nearest centroid classifier. Test which metric has the best cross validated performance: Euclidean, Cosine, or Manhattan. In `scikit-learn` you can see the documentation for NearestCentroid here: 
- http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestCentroid.html#sklearn.neighbors.NearestCentroid

and for supported distance metrics here:
- http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.distance_metrics.html#sklearn.metrics.pairwise.distance_metrics

In [14]:
from sklearn.neighbors.nearest_centroid import NearestCentroid
best_acc = 0
best_metrix = 'l1'
# the parameters for the nearest centroid metric to test are:
#    l1, l2, and cosine (all are optimized)
for d in ['l1','l2','cosine']:
    clf = NearestCentroid(metric=d)
    clf.fit(X_train, y_train)
    y_hat = clf.predict(X_test)
    acc = accuracy_score(y_test,y_hat)
    if acc> best_acc:
        best_acc = acc
        best_metrix=d
    print('The acuracy for ', d, ' is :', acc)
print('The best distance metric is: ', best_metrix)


/Users/alsabay/anaconda/lib/python3.6/site-packages/sklearn/neighbors/nearest_centroid.py:138: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "


The acuracy for  l1  is : 0.307406752696
The acuracy for  l2  is : 0.400742442991
The acuracy for  cosine  is : 0.478345412763
The best distance metric is:  cosine


___
<a id="naive"></a>
<a href="#top">Back to Top</a>
## Naive Bayes Classification
Now let's look at the use of the Naive Bayes classifier. The 20 newsgroups dataset has 20 classes and about 130,000 features per instance. Recall that the Naive Bayes classifer calculates a posterior distribution for each possible class. Each posterior distribution is a multiplication of many conditional distributions: 

$${\arg \max}_{j} \left(p(class=j)\prod_{i} p(attribute=i|class=j) \right)$$

where $p(class=j)$ is the prior and $p(attribute=i|class=j)$ is the conditional probability.

**Question for part 5**: With this many classes and features, how many different conditional probabilities need to be parameterized? How many priors need to be parameterized?

Enter you answer here:

There will be 2602140 conditional probabilities need to be parameterized. There will be 20 priors to be parametarized. 


20*130000 = 2602140


___
## Naive Bayes in Scikit-learn
Scikit has several implementations of the Naive Bayes classifier: `GaussianNB`, `MultinomialNB`, and `BernoulliNB`. Look at the documentation here: http://scikit-learn.org/stable/modules/naive_bayes.html Take a look at each implementation and then answer this question: 

**Questions for part 6**: 
- If the instances contain mostly continuous attributes, would it be better to use Gaussian Naive Bayes, Multinomial Naive Bayes, or Bernoulli? And Why? 
- What if the data is sparse, does this change your answer? Why or Why not?

#### Enter you answer here:

* For continous attributes, Gaussian Naive Bayes is a good choice because it's based on normal distribution. Multinomial Naive Bayes is used mainly in cases where features are discrete counts and Gaussian Naive Bayes is used where features are continuos.
* For sparse data, Multinomial or Bernoulli Naive Bayes will be better because the features simpler to evaluate (counting and binary) and will be more computationally efficient.
___

## Naive Bayes Comparison
For the final section of this notebook let's compare the performance of Naive Bayes for document classification. Look at the parameters for `MultinomialNB`, and `BernoulliNB` (especially `alpha` and `binarize`). 

**Exercise for part 7**: Using the example code below, change the parameters for each classifier and see how accurate you can make the classifiers on the test set. 

**Question for part 7**: Why are these implementations so fast to train? What does the `'alpha'` value control in these models (*i.e.*, how does it change the parameterizations)? 

In [29]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB


clf_mnb = MultinomialNB(alpha=1.0)
y_hat_mnb = clf_mnb.fit(X_train, y_train).predict(X_test)
clf_bnb = BernoulliNB(alpha=1.0, binarize=0.0)
y_hat_bnb = clf_bnb.fit(X_train, y_train).predict(X_test)
# fill in your code here
acc_mnb = accuracy_score(y_test,y_hat_mnb)
acc_bnb = accuracy_score(y_test,y_hat_bnb)
print(acc_mnb, acc_bnb)

clf_mnb = MultinomialNB(alpha=0.5)
y_hat_mnb = clf_mnb.fit(X_train, y_train).predict(X_test)
clf_bnb = BernoulliNB(alpha=0.5, binarize=0.0)
y_hat_bnb = clf_bnb.fit(X_train, y_train).predict(X_test)
# fill in your code here
acc_mnb = accuracy_score(y_test,y_hat_mnb)
acc_bnb = accuracy_score(y_test,y_hat_bnb)
print(acc_mnb, acc_bnb)

clf_mnb = MultinomialNB(alpha=0.1)
y_hat_mnb = clf_mnb.fit(X_train, y_train).predict(X_test)
clf_bnb = BernoulliNB(alpha=0.1, binarize=0.0)
y_hat_bnb = clf_bnb.fit(X_train, y_train).predict(X_test)
# fill in your code here
acc_mnb = accuracy_score(y_test,y_hat_mnb)
acc_bnb = accuracy_score(y_test,y_hat_bnb)
print(acc_mnb, acc_bnb)

clf_mnb = MultinomialNB(alpha=0.001)
y_hat_mnb = clf_mnb.fit(X_train, y_train).predict(X_test)
clf_bnb = BernoulliNB(alpha=0.001, binarize=0.002)
y_hat_bnb = clf_bnb.fit(X_train, y_train).predict(X_test)
# fill in your code here
acc_mnb = accuracy_score(y_test,y_hat_mnb)
acc_bnb = accuracy_score(y_test,y_hat_bnb)
print(acc_mnb, acc_bnb)
print('')
print('The last iteration using alpha of 0.001 and binarize = 0.002 is most accurate')
print('')
p = 'These classifiers are fast because there is no iteration required. For Bernoulli, it decides on a yes or no basis for example whether a word is in the document on not. For Multinomial, it is just a question of further counting how many times that word occurred. The parameters of the model for a Naive Bayes classifier are priors and conditional probabilities. This mainly involve just counting and division operations. On modern computers this is a very fast process. And hence naive bayes is overall fast.'
p2 = 'Multinomial Naive Bayes and Bernoulli Naive Bayes have a parameter alpha that controls the model complexity. The value of alpha adds alpha virtual points that have positive values for all the features. This helps in smoothing of the statistics. The large alpha means more smoothing, resulting in less complex model. Tunning the alpha usually improves the accuracy of the fit.'
print('These classifiers are so fast because...', p  )
print('')
print('The alpha values control...', p2)

0.701785398621 0.615167049673
0.758882800071 0.682517235284
0.830475517059 0.784161216192
0.890931589182 0.843026339049

The last iteration using alpha of 0.001 and binarize = 0.002 is most accurate

These classifiers are so fast because... These classifiers are fast because there is no iteration required. For Bernoulli, it decides on a yes or no basis for example whether a word is in the document on not. For Multinomial, it is just a question of further counting how many times that word occurred. The parameters of the model for a Naive Bayes classifier are priors and conditional probabilities. This mainly involve just counting and division operations. On modern computers this is a very fast process. And hence naive bayes is overall fast.

The alpha values control... Multinomial Naive Bayes and Bernoulli Naive Bayes have a parameter alpha that controls the model complexity. The value of alpha adds alpha virtual points that have positive values for all the features. This helps in smooth

These classifiers are fast because there's no iteration required. For Bernoulli, it decides on a yes or no basis for example whether a word is in the document on not. For Multinomial, it is just a question of further counting how many times that word occurred. The parameters of the model for a Naive Bayes classifier are priors and conditional probabilities. This mainly involve just counting and division operations. On modern computers this is a very fast process. And hence naive bayes is overall fast.

________________________________________________________________________________________________________

That's all! Please **upload your rendered notebook to blackboard** and please include **team member names** in the notebook submission.